subsys/net/lib/sockets/sockets.c

```diff
@@ -209,45 +210,92 @@ int zsock_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
 	return POINTER_TO_INT(ctx);
 }
 
+static struct net_pkt *prepare_tx_pkt(struct net_context *ctx,
+				      const void *buf, size_t len,
+				      s32_t timeout)
+{
+	struct net_pkt *pkt = NULL;
+	size_t max_len = net_if_get_mtu(net_context_get_iface(ctx));
+
+	pkt = net_pkt_get_tx(ctx, timeout);
+	if (!pkt) {
+		return NULL;
+	}
+
+	/* Make sure we don't send more data in one packet than
+	 * MTU allows. Optimize for number of branches in the code.
+	 */
+	max_len -= NET_IPV4TCPH_LEN;
+	if (net_context_get_family(ctx) != AF_INET) {
+		max_len -= NET_IPV6TCPH_LEN - NET_IPV4TCPH_LEN;
+	}
+
+	if (len > max_len) {
+		len = max_len;
+	}
+
+	len = net_pkt_append(pkt, len, buf, timeout);
+	if (!len) {
+		net_pkt_unref(pkt);
+		return NULL;
+	}
+
+	return pkt;
+}
+
 ssize_t zsock_send(int sock, const void *buf, size_t len, int flags)
 {
-	ARG_UNUSED(flags);
 	int err;
 	struct net_pkt *send_pkt;
 	s32_t timeout = K_FOREVER;
 	struct net_context *ctx = INT_TO_POINTER(sock);
-	size_t max_len = net_if_get_mtu(net_context_get_iface(ctx));
+
+	ARG_UNUSED(flags);
 
 	if (sock_is_nonblock(ctx)) {
 		timeout = K_NO_WAIT;
 	}
 
-	send_pkt = net_pkt_get_tx(ctx, timeout);
+	send_pkt = prepare_tx_pkt(ctx, buf, len, timeout);
 	if (!send_pkt) {
 		errno = EAGAIN;
 		return -1;
 	}
 
-	/* Make sure we don't send more data in one packet than
-	 * MTU allows. Optimize for number of branches in the code.
-	 */
-	max_len -= NET_IPV4TCPH_LEN;
-	if (net_context_get_family(ctx) != AF_INET) {
-		max_len -= NET_IPV6TCPH_LEN - NET_IPV4TCPH_LEN;
+	err = net_context_send(send_pkt, /*cb*/NULL, timeout, NULL, NULL);
+	if (err < 0) {
+		net_pkt_unref(send_pkt);
+		errno = -err;
+		return -1;
 	}
 
-	if (len > max_len) {
-		len = max_len;
+	return len;
+}
+
+ssize_t zsock_sendto(int sock, const void *buf, size_t len, int flags,
+		     const struct sockaddr *dest_addr, socklen_t addrlen)
+{
+	int err;
+	struct net_pkt *send_pkt;
+	s32_t timeout = K_FOREVER;
+	struct net_context *ctx = INT_TO_POINTER(sock);
+
+	ARG_UNUSED(flags);
+
+	if (sock_is_nonblock(ctx)) {
+		timeout = K_NO_WAIT;
 	}
 
-	len = net_pkt_append(send_pkt, len, buf, timeout);
-	if (!len) {
-		net_pkt_unref(send_pkt);
+	send_pkt = prepare_tx_pkt(ctx, buf, len, timeout);
+	if (!send_pkt) {
 		errno = EAGAIN;
 		return -1;
 	}
 
-	err = net_context_send(send_pkt, /*cb*/NULL, timeout, NULL, NULL);
+	SET_ERRNO(net_context_recv(ctx, zsock_received_cb, K_NO_WAIT, NULL));
```

Comment:
Why is this line here? (And not in zsock_send()? zsock_send() and zsock_sendto() should behave the same, and as noted previously, the way to ensure that is make one tail-call another).

Reply:
My assumption is, for TCP case, zsock_connect() will be called first and callback will be registered at that time.  So I only add it in zsock_sendto().

If we make zsock_send() and zsock_send() in adhoc way, any idea to prevent the registering the callback twice? Is there a proper way in net context to check if the callback has been registered?

Reply:
I see. Yep, there're some non-obvious points which only come to attention when you dig deeper into things. (Which may hint that there should be a code comment about it!)

> If we make zsock_send() and zsock_send() in adhoc way, any idea to prevent the registering the callback twice? Is there a proper way in net context to check if the callback has been registered?

Well, so you already register a callback here multiple times - each call to sendto() will lead to callback being registered. If that works, then testing for UDP and setting it in send() should also work.

But overall, that's a similar problem to bind_default(). I'd say, making send() to tail-cal to sendto() is more priority than avoiding duplicate setting if recv callback. Let's try to optimize it as we can now (i.e. set it in send*() only for UDP case), and I already accept that we'll need a much later pass over the socket code to see how we can optimize things.